plugins/aggregators/derivative/derivative.go

```diff
@@ -0,0 +1,225 @@
+	## measurement. This field is assumed to be monotonously increasing. This
+	## feature is used by specifying a *variable*.
+	## Make sure the specified variable is not filtered and exists in the metrics 
+	## passed to this aggregator!
+	# variable = ""
+	##
+	## When using a field as the derivation parameter the name of that field will
+	## be used for the resulting derivative, e.g. *fieldname_by_parameter*.
+	##
+	## Note, that the calculation is based on the actual timestamp of the
+	## measurements. When there is only one measurement during that period, the
+	## measurement will be rolled over to the next period. The maximum number of
+	## such roll-overs can be configured with a default of 10.
+	# max_roll_over = 10
+	##
+`
+
+func (d *Derivative) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *Derivative) Description() string {
+	return "Calculates a derivative for every field."
+}
+
+func (d *Derivative) Add(in telegraf.Metric) {
+	id := in.HashID()
+	if current, ok := d.cache[id]; !ok {
+		// hit an uncached metric, create caches for first time:
+		d.cache[id] = newAggregate(in)
+	} else {
+		if current.first.time.After(in.Time()) {
+			current.first = newEvent(in)
+			current.rollOver = 0
+		} else if current.first.time.Equal(in.Time()) {
+			upsertConvertedFields(in.Fields(), current.first.fields)
+			current.rollOver = 0
+		}
+		if current.last.time.Before(in.Time()) {
+			current.last = newEvent(in)
+			current.rollOver = 0
+		} else if current.last.time.Equal(in.Time()) {
+			upsertConvertedFields(in.Fields(), current.last.fields)
+			current.rollOver = 0
+		}
+	}
+}
+
+func newAggregate(in telegraf.Metric) *aggregate {
+	event := newEvent(in)
+	return &aggregate{
+		name:     in.Name(),
+		tags:     in.Tags(),
+		first:    event,
+		last:     event,
+		rollOver: 0,
+	}
+}
+
+func newEvent(in telegraf.Metric) *event {
+	return &event{
+		fields: extractConvertedFields(in),
+		time:   in.Time(),
+	}
+}
+
+func extractConvertedFields(in telegraf.Metric) map[string]float64 {
+	fields := make(map[string]float64, len(in.Fields()))
+	upsertConvertedFields(in.Fields(), fields)
+	return fields
+}
+
+func upsertConvertedFields(source map[string]interface{}, target map[string]float64) {
+	for k, v := range source {
+		if value, ok := convert(v); ok {
+			target[k] = value
+		}
+	}
+}
+
+func convert(in interface{}) (float64, bool) {
+	switch v := in.(type) {
+	case float64:
+		return v, true
+	case int64:
+		return float64(v), true
+	case uint64:
+		return float64(v), true
+	}
+	return 0, false
+}
+
+func (d *Derivative) Push(acc telegraf.Accumulator) {
+	for _, aggregate := range d.cache {
+		if aggregate.first == aggregate.last {
+			d.Log.Warnf("Same first and last event for %q, skipping.", aggregate.name)
+			continue
+		}
+		var denominator float64
+		if len(d.Variable) == 0 {
+			// If no derivation variable is known default to timestamp
+			denominator = aggregate.last.time.Sub(aggregate.first.time).Seconds()
```

Comment:
can you get rid of the else by making this the default value and flipping the condition?

Reply:
I do not know if I understand your suggestion. Wouldn't this mean I would always calculate the time differences, even when the are not needed?

Reply:
The idea of @ssoroka  is to do
```golang
denominator = aggregate.last.time.Sub(aggregate.first.time).Seconds()
if len(d.Variable) > 0 {
```
instead of the if/else construct. However this means that you, in the non-default-case, do a superfluous computation of the time difference. Don't know if this matters...

Reply:
This is how I understood @ssoroka. I just wanted to make sure, that I really should always calculate the time difference, even when it is not needed. I will make that change.